File: Algotithms/sorting-algorithms/Quick-sort.cpp
/* Quick sort algorithm in cpp:*/

#include "iostream"

using namespace std;

/* swap function:*/
void swap(int & a, int & b){
    int swap_var = a;
    a = b;
    b = swap_var;
    return;
}

/* function to print the array:*/
void printArray(int * array, ssize_t size){
    for (int i = 0; i < size; i ++){
        cout << array[i] << " ";
    }
    cout << endl;
    return;
}

/* function to rearrange the array and find the partition point: */
int partition(int * array, int low, int high){
    /* select rightmost element as pivot: */
    int pivot = array[high];
    /* pointer for greater element: */
    int i = low - 1;

    /* compare each element with the pivot: */

    for (int  j= low; j < high; j ++){
        if (array[j] <= pivot){
            /* swap this element with the pivot element pointed to by i: */
            i ++;
            swap(array[i], array[j]);
        }
    }

    /* swap pivot with greater element at i: */
    swap(array[i + 1], array[high]);

    /* return the partition point: */
    return ( i + 1 );
}

/* quicksort function: */
void quickSort(int * array, int low, int high){
    if (low < high){
        /* find the pivot elements such that elements smaller than 
        the pivot are on the left while those that are greater than 
        the pivot are on the right: */
        int pivot_point = partition(array, low, high);
        /* recursively call on the left of the pivot: */

        quickSort(array, low, pivot_point - 1);
        quickSort(array, pivot_point + 1, high);
    }
    return;
}

int main(){
    int data[] = {-2, -45, 0, 11, -9};
    ssize_t size = sizeof(data) / sizeof(*data);
    cout <<"Array before sorting-> " << endl;
    printArray(data, size);
    quickSort(data, 0, size - 1);
    cout << "Sorted Array in ascending order-> " << endl;
    printArray(data, size);
    return 0;
}

File: Algorithms/sorting-algorithms/Quick-sort.py
""" Quick sort algorithm in python: """

def partition(array, low, high):
    """ choose the rightmost element as the pivot:"""
    pivot = array[high]
    """pointer for greater element:"""

    i = low - 1

    """ traverse through all elements comparing each with pivot.
    """
    for j in range(low, high):
        if array[j] <= pivot:
            """if element smaller than pivot is found, swap it with 
            greater element pointed to by i"""

            i = i + 1
            array[i], array[j] = array[j], array[i]
    """swap the pivot element with the greater element specified by i"""
    array[i + 1], array[high] = array[high], array[i + 1]

    """return the position from where the partition is done:"""
    return i + 1


def quickSort(array, low, high):
    if low < high:
        """find the pivot element:"""
        pi = partition(array, low, high)
        
        """recursive call on the left of the pivot:"""
        quickSort(array, low, pi - 1)
        quickSort(array, pi + 1, high)
        
def printArray(array):
    for i in range(len(array)):
        print(array[i], end=" ")
    print()

if __name__ == "__main__":
    data = [-2, 45, 0, 11, -9]
    size = len(data)

    quickSort(data, 0, size - 1)
    printArray(data)